File: src/BehaviorDiff.Java.Agent/src/main/java/io/behaviordiff/agent/RedactionPolicy.java
package io.behaviordiff.agent;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;

final class RedactionPolicy {
    private static final List<String> DEFAULT_NAMES = Arrays.asList(
        "password", "token", "secret", "key", "ssn", "email", "auth", "credential");
    private static final Pattern JWT = Pattern.compile(
        "\\beyJ[A-Za-z0-9_-]{10,}\\.[A-Za-z0-9_-]{10,}\\.[A-Za-z0-9_-]{10,}\\b");
    private static final Pattern AWS = Pattern.compile("\\b(?:AKIA|ASIA)[0-9A-Z]{16}\\b");
    private static final Pattern PEM = Pattern.compile("-----BEGIN [A-Z0-9 ]*(?:PRIVATE KEY|CERTIFICATE)-----");
    private static final Pattern BASE64 = Pattern.compile(
        "(?:^|[^A-Za-z0-9+/])(?:[A-Za-z0-9+/]{40,}={0,2})(?:$|[^A-Za-z0-9+/=])");

    private final List<String> names;
    private final List<String> types;
    private final List<String> paths;

    private RedactionPolicy(List<String> names, List<String> types, List<String> paths) {
        this.names = names;
        this.types = types;
        this.paths = paths;
    }

    static RedactionPolicy fromEnvironment() {
        List<String> names = new ArrayList<>(DEFAULT_NAMES);
        names.addAll(readList(System.getenv("BEHAVIORDIFF_REDACT_NAMES")));
        return new RedactionPolicy(names, readList(System.getenv("BEHAVIORDIFF_REDACT_TYPES")),
            readList(System.getenv("BEHAVIORDIFF_REDACT_PATHS")));
    }

    boolean sensitiveName(String value) {
        String normalized = value.toLowerCase(Locale.ROOT);
        return names.stream().anyMatch(normalized::contains);
    }

    boolean digestOnlyType(Class<?> type) {
        String name = type.getName().toLowerCase(Locale.ROOT);
        return types.stream().anyMatch(prefix -> name.equals(prefix) || name.startsWith(prefix + "."));
    }

    boolean digestOnlyPath(String path) {
        if (path == null) return false;
        String normalized = normalizePath(path);
        return paths.stream().anyMatch(prefix -> normalized.equals(prefix)
            || normalized.startsWith(prefix + "/")
            || normalized.endsWith("/" + prefix)
            || normalized.contains("/" + prefix + "/"));
    }

    static boolean credentialContent(String value) {
        return JWT.matcher(value).find() || AWS.matcher(value).find() || PEM.matcher(value).find()
            || BASE64.matcher(value).find();
    }

    private static List<String> readList(String value) {
        List<String> result = new ArrayList<>();
        if (value == null) return result;
        for (String item : value.split("[;,]")) {
            String normalized = item.trim().toLowerCase(Locale.ROOT);
            if (!normalized.isEmpty() && !result.contains(normalized)) result.add(normalizePath(normalized));
        }
        return result;
    }

    private static String normalizePath(String value) {
        String normalized = value.replace('\\', '/');
        while (normalized.startsWith("/")) normalized = normalized.substring(1);
        while (normalized.endsWith("/")) normalized = normalized.substring(0, normalized.length() - 1);
        return normalized;
    }
}

File: src/RealDiff.Tracer/RedactionPolicy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace RealDiff.Tracer
{
    internal sealed class RedactionPolicy
    {
        private static readonly string[] DefaultNames =
        {
            "password", "token", "secret", "key", "ssn", "email", "auth", "credential",
        };

        private static readonly Regex Jwt = new(
            @"\beyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\b",
            RegexOptions.Compiled | RegexOptions.CultureInvariant);
        private static readonly Regex AwsKey = new(
            @"\b(?:AKIA|ASIA)[0-9A-Z]{16}\b",
            RegexOptions.Compiled | RegexOptions.CultureInvariant);
        private static readonly Regex Pem = new(
            @"-----BEGIN [A-Z0-9 ]*(?:PRIVATE KEY|CERTIFICATE)-----",
            RegexOptions.Compiled | RegexOptions.CultureInvariant);
        private static readonly Regex LongBase64 = new(
            @"(?:^|[^A-Za-z0-9+/])(?:[A-Za-z0-9+/]{40,}={0,2})(?:$|[^A-Za-z0-9+/=])",
            RegexOptions.Compiled | RegexOptions.CultureInvariant);

        internal RedactionPolicy(
            IReadOnlyList<string> sensitiveNames,
            IReadOnlyList<string> digestOnlyTypes,
            IReadOnlyList<string> digestOnlyPaths)
        {
            SensitiveNames = sensitiveNames;
            DigestOnlyTypes = digestOnlyTypes;
            DigestOnlyPaths = digestOnlyPaths;
        }

        internal IReadOnlyList<string> SensitiveNames { get; }

        internal IReadOnlyList<string> DigestOnlyTypes { get; }

        internal IReadOnlyList<string> DigestOnlyPaths { get; }

        internal static RedactionPolicy FromEnvironment() => new(
            MergeDefaults(Environment.GetEnvironmentVariable(TracerOptions.RedactNamesVariable)),
            ReadList(Environment.GetEnvironmentVariable(TracerOptions.RedactTypesVariable)),
            ReadList(Environment.GetEnvironmentVariable(TracerOptions.RedactPathsVariable))
                .Select(NormalizePath).ToArray());

        internal bool IsSensitiveName(string? name) => name != null
            && SensitiveNames.Any(pattern => name.IndexOf(pattern, StringComparison.OrdinalIgnoreCase) >= 0);

        internal bool IsDigestOnlyType(Type type)
        {
            string name = type.FullName ?? type.Name;
            return DigestOnlyTypes.Any(pattern => PrefixMatch(name, pattern));
        }

        internal bool IsDigestOnlyPath(string? path)
        {
            if (string.IsNullOrWhiteSpace(path))
            {
                return false;
            }

            string normalized = NormalizePath(path!);
            return DigestOnlyPaths.Any(pattern =>
                normalized == pattern
                || normalized.StartsWith(pattern + "/", StringComparison.OrdinalIgnoreCase)
                || normalized.EndsWith("/" + pattern, StringComparison.OrdinalIgnoreCase)
                || normalized.IndexOf("/" + pattern + "/", StringComparison.OrdinalIgnoreCase) >= 0);
        }

        internal static bool ContainsCredential(string value) => Jwt.IsMatch(value)
            || AwsKey.IsMatch(value)
            || Pem.IsMatch(value)
            || LongBase64.IsMatch(value);

        private static IReadOnlyList<string> MergeDefaults(string? configured) => DefaultNames
            .Concat(ReadList(configured))
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToArray();

        private static IReadOnlyList<string> ReadList(string? value) => string.IsNullOrWhiteSpace(value)
            ? Array.Empty<string>()
            : value!.Split(new[] { ';', ',' }, StringSplitOptions.RemoveEmptyEntries)
                .Select(item => item.Trim())
                .Where(item => item.Length > 0)
                .Distinct(StringComparer.OrdinalIgnoreCase)
                .ToArray();

        private static string NormalizePath(string path) => path.Replace('\\', '/').Trim().Trim('/');

        private static bool PrefixMatch(string value, string prefix) =>
            value.Equals(prefix, StringComparison.OrdinalIgnoreCase)
            || value.StartsWith(prefix + ".", StringComparison.OrdinalIgnoreCase)
            || value.StartsWith(prefix + "`", StringComparison.OrdinalIgnoreCase);
    }
}